# DENTIST R and cpp cpmparison


## Overview

DENTIST code is implemented in cpp `https://github.com/Yves-CHEN/DENTIST`. To incorporate it in the pecotmr package an R version is developed. 

Here we will test if the R version code have the same result of cpp version.

## Important note

DENTIST does not have a easy way to compile it in the system. Their paths in the makefile is quite messy and we need to install all dependencies and re-wrote the make file, which is quite time consuming. 

However, DENTIST provided an excutable file that can do the same things, but we cannot modify it so that we can only have input and outputs. Given that DENTIST involves randomness in splitting the region into S1 and S2 two parts and iterated this multiple times, so the numerical value of different attempts might be different. So here our strategy is to run DENTIST 10 times in each system, take the average for numerical things and compare if they are similar enough.

## Input

DENTIST the original program needs PLINK file as input, and also sumstat in COJO format (including columns `SNP A1 A2 freq b se p N`). Here we use the first round RSS_QC result as reference, found a region have outliers after allele QC. We use this region to verify the similarity of result.

+ PLINK file: `/home/hs3393/RSS_QC/pecotmr/data/RSS_QC_MWE/PLINK_input_MWE`

+ Sumstat: `/home/hs3393/RSS_QC/pecotmr/data/RSS_QC_MWE/Sumstat_MAF_filtered.txt`


Notes about inputs: DENTIST only allow region with 2000+ variables, also, should not include variants that have maf = 0 in PLINK. So here the plink file is already filtered by threshold 0.01.

## Details

### 1. MWE data preparation

For DENTIST in github https://github.com/Yves-CHEN/DENTIST version 1.3.0.0, it requires PLINK data, so here I prepared it.

In [6]:
library(tidyverse)
library(susieR)
library(plink2R)
library(pecotmr)
library(data.table)

# Original sumstat is from Bellenguez GWAS data. 
# I found a region that has some signal and outliers so perfect for MWE.
# Keep ~ 4000 variants because it's demanded by DENTIST

test_data_path = "/home/hs3393/RSS_QC/pecotmr/data/RSS_QC_MWE/Original_sumstat.tsv"
snp_data <- fread(test_data_path)

# get variant list to extract from PLINK data
snp_data %>% mutate(variant_id = paste0("chr", chrom, ":", pos, "_", A2, "_", A1)) %>% pull(variant_id) %>% 
    write.table("/home/hs3393/RSS_QC/pecotmr/data/RSS_QC_MWE/MWE_snplist.txt", row.names = F, quote = F)


Attaching package: ‘data.table’


The following objects are masked from ‘package:lubridate’:

    hour, isoweek, mday, minute, month, quarter, second, wday, week,
    yday, year


The following objects are masked from ‘package:dplyr’:

    between, first, last


The following object is masked from ‘package:purrr’:

    transpose




In [ ]:
plink --bfile /mnt/vast/hpc/csg/FunGen_xQTL/ROSMAP/Genotype/geno_by_chrom/ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.1 \ 
    --extract /home/hs3393/RSS_QC/pecotmr/data/RSS_QC_MWE/MWE_snplist.txt \
    --make-bed --maf 0.01 --geno 0.01 --out ~/RSS_QC/pecotmr/data/RSS_QC_MWE/PLINK_input_MWE

In [21]:
genotype = read_plink("~/RSS_QC/pecotmr/data/RSS_QC_MWE/PLINK_input_MWE")

# filter out columns that contain NA so that LD matrix can be computed without error
keep_index = which(!is.na(colSums(genotype$bed)))

names(keep_index) %>% 
    write.table("/home/hs3393/RSS_QC/pecotmr/data/RSS_QC_MWE/MWE_snplist.txt", row.names = F, quote = F)

snp_data %>% mutate(variant_id = paste0("chr", chrom, ":", pos, "_", A2, "_", A1)) %>%
    filter(variant_id %in% names(keep_index)) %>% mutate(pvalue = pnorm(beta / se)) %>%
    select(variant_id, A1, A2, maf, beta, se,  pvalue ,n_case) %>% rename(SNP = variant_id,
                                                                        freq = maf, b = beta,
                                                                        p = pvalue, N = n_case) %>%
    mutate(N = 1153) %>%
    write_tsv("~/RSS_QC/pecotmr/data/RSS_QC_MWE/Sumstat_MAF_filtered.txt")

In [ ]:
# rerun plink to remove those variants containing NA
plink --bfile /mnt/vast/hpc/csg/FunGen_xQTL/ROSMAP/Genotype/geno_by_chrom/ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.1 \ 
    --extract /home/hs3393/RSS_QC/pecotmr/data/RSS_QC_MWE/MWE_snplist.txt \
    --make-bed --maf 0.01 --geno 0.01 --out ~/RSS_QC/pecotmr/data/RSS_QC_MWE/PLINK_input_MWE

In [ ]:
head -n 6 /home/hs3393/RSS_QC/pecotmr/data/RSS_QC_MWE/MWE_snplist.txt >> /home/hs3393/RSS_QC/pecotmr/data/RSS_QC_MWE/MWE_top5_snplist.txt
plink --bfile /mnt/vast/hpc/csg/FunGen_xQTL/ROSMAP/Genotype/geno_by_chrom/ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.1 --extract /home/hs3393/RSS_QC/pecotmr/data/RSS_QC_MWE/MWE_top5_snplist.txt --make-bed --maf 0.01 --geno 0.01 --out /home/hs3393/RSS_QC/pecotmr/data/RSS_QC_MWE/PLINK_top5_MWE

In [39]:
# calculate LD matrix to run it in rCPP interface 
#sumstat = read_delim("/home/hs3393/RSS_QC/pecotmr/data/RSS_QC_MWE/Sumstat_MAF_filtered.txt")
#geno = read_plink("/home/hs3393//RSS_QC/pecotmr/data/RSS_QC_MWE/PLINK_input_MWE")$bed
#LD = as.matrix(cor(geno))
#write.table(LD, "/home/hs3393//RSS_QC/pecotmr/data/RSS_QC_MWE/LD_MWE.tsv", sep = "\t", quote = FALSE, row.names = TRUE, col.names = TRUE)

#########FINISHED############
sumstat = read_delim("/home/hs3393/RSS_QC/pecotmr/data/RSS_QC_MWE/Sumstat_MAF_filtered.txt")
LD = fread("/home/hs3393//RSS_QC/pecotmr/data/RSS_QC_MWE/LD_MWE.tsv")[,-1]
LD = as.matrix(LD)

Rows: 4123 Columns: 8
── Column specification ────────────────────────────────────────────────────────
Delimiter: "\t"
chr (3): SNP, A1, A2
dbl (5): freq, b, se, p, N

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 4123 Columns: 8
── Column specification ────────────────────────────────────────────────────────
Delimiter: "\t"
chr (3): SNP, A1, A2
dbl (5): freq, b, se, p, N

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Warning message in fread("/home/hs3393//RSS_QC/pecotmr/data/RSS_QC_MWE/LD_MWE.tsv"):
“Detected 4123 column names but the data has 4124 columns (i.e. invalid file). Added 1 extra default column name for the first column which is guessed to be row names or an index. Use setnames() afterwards if this guess is not correct, or fix the file write command that 

### 2. DENTIST  -- github compiled version implementation

In [ ]:
# DENTIST -- github compiled version

~/RSS_QC/DENTIST/DENTIST_1.3.0.0  --bfile  /home/hs3393//RSS_QC/pecotmr/data/RSS_QC_MWE/PLINK_input_MWE --gwas-summary /home/hs3393/RSS_QC/pecotmr/data/RSS_QC_MWE/Sumstat_MAF_filtered.txt --out DENTIST_result

In [20]:
wc -l ~/RSS_QC/DENTIST/DENTIST_result.DENTIST.short.txt

63 /home/hs3393/RSS_QC/DENTIST/DENTIST_result.DENTIST.short.txt


Here the row number of *.short.txt shows how many outliers are identified in this region and should be removed.

So for DENTIST compiled version, the **number of outlier is 63**. 

### 3. DENTIST -- Rcpp version

In [35]:
sumstat = read_delim("/home/hs3393/RSS_QC/pecotmr/data/RSS_QC_MWE/Sumstat_MAF_filtered.txt") %>% mutate(z = b / se)
LD = fread("/home/hs3393//RSS_QC/pecotmr/data/RSS_QC_MWE/LD_MWE.tsv")[,-1]
LD = as.matrix(LD)
dentist_result = dentist(zScore = sumstat$z, LD = LD, nSample = 1153)

Rows: 4123 Columns: 8
── Column specification ────────────────────────────────────────────────────────
Delimiter: "\t"
chr (3): SNP, A1, A2
dbl (5): freq, b, se, p, N

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Warning message in fread("/home/hs3393//RSS_QC/pecotmr/data/RSS_QC_MWE/LD_MWE.tsv"):
“Detected 4123 column names but the data has 4124 columns (i.e. invalid file). Added 1 extra default column name for the first column which is guessed to be row names or an index. Use setnames() afterwards if this guess is not correct, or fix the file write command that created the file to create a valid file.”


In [36]:
dentist_result %>% filter(is_problematic > 0) %>% nrow()

[1] 1053

So for DENTIST RCPP version, the **number of outlier is 1053**. The results looks quite different... Take a look at the LD.

In [52]:
library(readr)
library(dplyr, warn.conflicts = FALSE)
library(data.table)
library(RcppArmadillo)
library(Rcpp)
source("/home/rd2972/software/pecotmr/R/run_dentist.R")
source("/home/rd2972/software/pecotmr/R/RcppExports.R")
sourceCpp("/home/hs3393/RSS_QC/pecotmr/src/dentist.cpp")
sourceCpp("/home/rd2972/software/pecotmr/src/RcppExports.cpp")
#dentist_LD_result = fread("/home/rd2972/private_data/20240300_rss_qc_imputation/DENTIST/output_LD.txt")

Warning message in sourceCpp("/home/rd2972/software/pecotmr/src/RcppExports.cpp"):
“No Rcpp::export attributes or RCPP_MODULE declarations found in source”


In [55]:
LD

chr1:206011748_A_G chr1:206012565_A_G chr1:206012721_C_G
 [1,]  1.00000000         0.98500629         0.98215134       
 [2,]  0.98500629         1.00000000         0.99703144       
 [3,]  0.98215134         0.99703144         1.00000000       
 [4,]  0.97722319         0.98655338         0.98367739       
 [5,] -0.16241358        -0.16597212        -0.16673209       
 [6,]  0.98013632         0.98062367         0.97479826       
 [7,]  0.94805046         0.94822880         0.94244220       
 [8,]  0.94283218         0.95174415         0.94886559       
 [9,]  0.31266017         0.30599136         0.30414100       
[10,]  0.41408706         0.40531837         0.40288333       
[11,]  0.58251096         0.57531861         0.57433986       
[12,]  0.28935799         0.28270539         0.28087510       
[13,]  0.19779936         0.20217400         0.21003039       
[14,]  0.18323232         0.19661504         0.21322948       
[15,]  0.16698877         0.19635348         0.21189003       
[16,]  0.18697656         0.20005226         0.21642110       
[17,]  0.23869597         0.24286537         0.24135066       
[18,]  0.19243985         0.19602804         0.19463296       
[19,]  0.18697656         0.20005226         0.21642110       
[20,]  0.19243985         0.19602804         0.19463296       
[21,]  0.18697656         0.20005226         0.21642110       
[22,]  0.20366795         0.19873534         0.19738577       
[23,]  0.18697656         0.20005226         0.21642110       
[24,]  0.18323232         0.19661504         0.21322948       
[25,]  0.16789131         0.19809598         0.21403379       
[26,] -0.07010229        -0.07143634        -0.07171487       
[27,]  0.55363363         0.54699978         0.54499157       
[28,]  0.16789131         0.19809598         0.21403379       
[29,]  0.19646769         0.19169692         0.19039200       
[30,]  0.17285144         0.20344025         0.21957233       
[31,] ⋮                  ⋮                  ⋮                 
[32,] -0.0091603060      -0.0126553405      -0.013505590      
[33,] -0.0093453696      -0.0127679696      -0.013600339      
[34,]  0.0349953546       0.0286634938       0.027087489      
[35,] -0.0091603060      -0.0126553405      -0.013505590      
[36,] -0.0091603060      -0.0126553405      -0.013505590      
[37,]  0.0295317561       0.0159497229       0.017325303      
[38,]  0.0124788325       0.0175219957       0.013545644      
[39,]  0.0238330715       0.0101161669       0.011426403      
[40,] -0.0093453696      -0.0127679696      -0.013600339      
[41,] -0.0093453696      -0.0127679696      -0.013600339      
[42,]  0.0103892240      -0.0007561224       0.003150928      
[43,] -0.0063708208      -0.0046928718      -0.005669176      
[44,] -0.0125115125      -0.0094938946      -0.010831264      
[45,]  0.0042552123      -0.0014634494       0.003336064      
[46,] -0.0069343675      -0.0104613643      -0.011320984      
[47,] -0.0069343675      -0.0104613643      -0.011320984      
[48,] -0.0069343675      -0.0104613643      -0.011320984      
[49,]  0.0005898472      -0.0014138337      -0.001905311      
[50,] -0.0069343675      -0.0104613643      -0.011320984      
[51,] -0.0170794133      -0.0157594164      -0.013559374      
[52,] -0.0219637636      -0.0242452626      -0.024789252      
[53,] -0.0193532592      -0.0216382623      -0.022184913      
[54,] -0.0138176149      -0.0055390197      -0.004293321      
[55,] -0.0394362221      -0.0419444286      -0.042531913      
[56,]  0.0256967804       0.0225030342       0.021702552      
[57,] -0.0053929916      -0.0077023607      -0.008264625      
[58,] -0.0135394991      -0.0153967323      -0.015842559      
[59,] -0.0461785952      -0.0429995460      -0.043702051      
[60,] -0.0443735833      -0.0411619006      -0.041865722      
[61,] -0.0411658907      -0.0382699506      -0.039014423      
      chr1:206013858_A_G chr1:206014205_G_A chr1:206015713_G_A
 [1,]  0.97722319        -0.16241358 

In [54]:
result1 = fread("/home/rd2972/private_data/20240300_rss_qc_imputation/DENTIST/per_iteration/dentist_Rcpp/20240308_version/DENTIST_Rcpp_output_01.txt")

V1          V2          V3         V4          V5         V6         
1     0.985010    0.997030   -0.165970   0.980620    0.9517400  0.3059900 
2     0.977220    0.983680   -0.165530   0.990970    0.9623700  0.3093300 
3     0.948050    0.942440   -0.078334   0.964930    0.9912700  0.2999800 
4     0.414090    0.402880   -0.056151   0.414770    0.3990600 -0.0349970 
5     0.582510    0.574340   -0.182270   0.589240    0.6041700  0.1733100 
6     0.289360    0.280880   -0.048627   0.290170    0.2728200  0.9004400 
7     0.197800    0.210030   -0.066790   0.198570    0.1854400  0.6333200 
8     0.183230    0.213230   -0.064954   0.184010    0.1714600  0.5908600 
9     0.186980    0.216420   -0.066252   0.187760    0.1750000  0.6072600 
10    0.203670    0.197390   -0.065482   0.204390    0.1912600  0.6430300 
11    0.186980    0.216420   -0.066252   0.187760    0.1750000  0.6072600 
12    0.183230    0.213230   -0.064954   0.184010    0.1714600  0.5908600 
13   -0.070102   -0.071715    0.375760  -0.068439   -0.0751770 -0.0351290 
14    0.553630    0.544990    0.448120   0.560930    0.5796500 -0.0552040 
15    0.196470    0.190390   -0.062798   0.197180    0.1844600  0.6368800 
16    0.172850    0.219570   -0.067530   0.191460    0.1784800  0.5745100 
17    0.183230    0.213230   -0.064954   0.184010    0.1714600  0.5908600 
18   -0.106950   -0.109250    0.532680  -0.104490   -0.1143100 -0.0335180 
19    0.177960    0.216420   -0.066252   0.187760    0.1750000  0.5825300 
20    0.177960    0.216420   -0.066252   0.187760    0.1750000  0.5825300 
21    0.177960    0.216420   -0.066252   0.187760    0.1750000  0.5825300 
22    0.177960    0.216420   -0.066252   0.187760    0.1750000  0.5825300 
23    0.195780    0.216480   -0.069344   0.205710    0.1921800  0.5899400 
24    0.201390    0.213280   -0.068077   0.202170    0.1888400  0.5985800 
25    0.195780    0.216480   -0.069344   0.205710    0.1921800  0.5899400 
26   -0.083472   -0.084728    0.360640  -0.081827   -0.0878720 -0.0334980 
27   -0.042468   -0.038344   -0.044900  -0.040446   -0.0082962 -0.0067722 
28    0.600050    0.596750    0.105730   0.609540    0.6185400  0.1349300 
29    0.591750    0.588410    0.089708   0.606890    0.5725100  0.1403300 
30    0.289490    0.297330    0.121350   0.297090    0.3109000 -0.0820300 
⋮    ⋮           ⋮           ⋮          ⋮           ⋮          ⋮          
2033 -0.02617300 -0.02911200 -0.0189420 -0.02436800 -0.0338530  0.00194470
2034 -0.04218500 -0.04634000 -0.0304290 -0.03957000 -0.0421160 -0.03011500
2035 -0.03122200 -0.02890300  0.0761340 -0.02564500 -0.0247160 -0.01611800
2036 -0.04218500 -0.04634000 -0.0304290 -0.03957000 -0.0421160 -0.03011500
2037 -0.03366700 -0.03872600 -0.0175730 -0.03568000 -0.0272390 -0.05695300
2038 -0.03379400 -0.03149500  0.0744000 -0.02817000 -0.0273780 -0.01695400
2039 -0.03300700 -0.03069600  0.0717960 -0.02738600 -0.0265470 -0.01671100
2040 -0.04218500 -0.04634000 -0.0304290 -0.03957000 -0.0421160 -0.03011500
2041 -0.01060900 -0.00912980  0.0713670 -0.00950030 -0.0020405  0.00203220
2042  0.00359900  0.00034147 -0.0099817  0.00520660 -0.0044106  0.02998500
2043 -0.02505600 -0.02356400  0.0465750 -0.03736000 -0.0300180  0.00876280
2044  0.04118400  0.03997900  0.0567980  0.05053700  0.0530240  0.01178700
2045 -0.00320020 -0.00747060 -0.0277660 -0.00099723 -0.0138220 -0.01951900
2046 -0.00916030 -0.01350600 -0.0195010 -0.00684750 -0.0081100 -0.02283700
2047  0.03499500  0.02708700 -0.0135120  0.03858100  0.0237290 -0.01523100
2048  0.01247900  0.01354600 -0.0363520  0.01296000  0.0164580 -0.03235800
2049 -0.00934540 -0.01360000 -0.0154410 -0.00707610 -0.0078823 -0.02120100
2050 -0.00934540 -0.01360000 -0.0154410 -0.00707610 -0.0078823 -0.02120100
2051  0.01038900  0.00315090  0.0148820  0.00345350 -0.0059086  0.02258200
2052 -0.00637080 -0.00566920 -0.0110040 -0.00382770 -0.0074890 -0.00030993
2053 -0.00693440 -0.01132100 -0.0140630 -0.00462750 -0.0062140 -0.02362200
2054 -0.00693440 -0.01132100 -

ERROR: Error in parse(text = x, srcfile = src): <text>:1:150: unexpected input
1: result2 = fread("/home/rd2972/private_data/20240300_rss_qc_imputation/DENTIST/per_iteration/dentist_Rcpp/20240308_version/DENTIST_Rcpp_output_11.txt"）
                                                                                                                                                         ^


In [68]:
fread("/home/rd2972/private_data/20240300_rss_qc_imputation/DENTIST/per_iteration/dentist_Rcpp/20240308_version/DENTIST_Rcpp_output_11.txt")

V1          V2         V3          V4         V5         V6         
1    -0.0416310  -0.0414220  0.01065200 -0.0392200 -0.0086544  0.0083688 
2    -0.0271160  -0.0334570  0.01177400 -0.0349700 -0.0307070  0.0151990 
3    -0.0150950   0.0788840  0.01142900  0.0272150  0.0175620 -0.0107720 
4    -0.0747820   0.0207080 -0.00889540 -0.0468830 -0.0358610 -0.0107720 
5    -0.0360670  -0.0478700 -0.20013000  0.0109740  0.0022737  0.0112420 
6     0.0703750  -0.0192480 -0.32660000 -0.0708390 -0.0766960  0.0011983 
7    -0.0535730   0.1145300  0.02604700 -0.0591690 -0.0409050  0.0435680 
8     0.0502220  -0.0088200  0.03224900  0.0054500  0.0342550 -0.0157530 
9    -0.0266670   0.1907900  0.00052716 -0.0060007  0.0028529 -0.0313640 
10   -0.1110200   0.0991160  0.05197900 -0.0325070  0.0186360 -0.0107480 
11   -0.0620930   0.0238440 -0.03332400  0.2061300  0.2509700  0.0099152 
12   -0.0435060   0.1311800 -0.01249900 -0.0161920 -0.0095078  0.0075131 
13   -0.1080100  -0.0549360 -0.04525400  0.1313300  0.1553500  0.0522830 
14    0.4634500   0.0529270 -0.00797730 -0.0415880 -0.0193930  0.0015763 
15   -0.0585380  -0.0060310  0.01314700 -0.0723300 -0.0682250  0.0172470 
16   -0.0146590  -0.0083168 -0.04609200 -0.0134300 -0.0068800  0.0101600 
17   -0.2110400   0.0773250  0.02943500 -0.0410800 -0.0216050 -0.0677960 
18   -0.0089169   0.0480860  0.08568500 -0.0101710 -0.0044931 -0.0312540 
19    0.0848390  -0.0075151 -0.05802500 -0.0483550 -0.0674110  0.0458250 
20   -0.0523080  -0.0804590 -0.00627550 -0.0203970  0.0104000  0.0035698 
21   -0.0874020   0.0771190  0.02272100 -0.0545720 -0.0370940 -0.0206220 
22   -0.0759360  -0.0539590 -0.01180500 -0.0132170 -0.0042418 -0.0068939 
23   -0.0758820   0.0221990 -0.01047100 -0.0473340 -0.0362760 -0.0111500 
24    0.0040177   0.0616510 -0.06571300  0.0151280 -0.0180330  0.0279950 
25    0.1485700   0.2049800  0.01629600  0.1786300  0.1597100 -0.0364980 
26   -0.0717830   0.0535500  0.03190400  0.0052320 -0.0280410 -0.0276690 
27   -0.0986790  -0.0780640 -0.01001800 -0.0291310 -0.0022100 -0.0083727 
28   -0.0747820   0.0207080 -0.00889540 -0.0468830 -0.0358610 -0.0107720 
29   -0.0605400  -0.0154570 -0.06864900  0.1224800  0.1745800  0.0202970 
30   -0.0213690  -0.0765380  0.02778500 -0.0397940 -0.0349430  0.0075131 
⋮    ⋮           ⋮          ⋮           ⋮          ⋮          ⋮          
1373 -0.01258500 -0.0316150  0.8303100  -0.0127210 -0.0166410 -0.01222600
1374 -0.00376660  0.0664350 -0.0418870  -0.0340520 -0.0262410  0.01802000
1375  0.22779000 -0.0725130  0.0274420  -0.0134300 -0.0068800  0.04587400
1376  0.01047800  0.0182690  0.0077282  -0.0392490 -0.0093339  0.22524000
1377 -0.04025800 -0.0122760  0.1486400  -0.0090746 -0.0029125 -0.02310900
1378 -0.00610850  0.0295000 -0.0196250   0.0253530  0.0261280 -0.10575000
1379  0.00689610 -0.0561890  0.0884470  -0.0330960 -0.0298700  0.00009349
1380 -0.00138380  0.0277380 -0.0632250   0.0217190  0.0351600 -0.03956100
1381  0.00572690 -0.0520220  0.0046386  -0.0424460 -0.0027947 -0.09633900
1382 -0.04025800 -0.0122760  0.1486400  -0.0090746 -0.0029125 -0.02310900
1383 -0.01995600 -0.0619300 -0.0026290   0.2643100 -0.0249760  0.21972000
1384  0.00953810  0.2425300  0.0082102  -0.0597500 -0.1487100  0.03330700
1385 -0.02398200  0.0164430 -0.0145690  -0.0168870 -0.0268830 -0.00943260
1386 -0.01176800  0.0400580  0.2929700   0.0056562  0.0107610 -0.00760110
1387 -0.19111000 -0.1600700 -0.0372970  -0.0511400 -0.0398550  0.05009800
1388 -0.06718200  0.1304600  0.0017562   0.7645000  0.9889000  0.00073003
1389  0.00366910 -0.0237620 -0.0394180   0.0158610  0.0273640  0.00682010
1390 -0.07236500  0.0681570  0.0290690  -0.0839440 -0.0640380  0.01653600
1391 -0.11127000  0.0747180  0.0016614  -0.0571470 -0.0715450 -0.01942400
1392  0.02822200  0.0058339  0.0145750   0.0525060  0.0202780 -0.03887100
1393  0.02378200 -0.0303060  0.0230070  -0.0454360 -0.0710510  0.01882300
1394 -0.09396500 -0.0094162 -0.2574700   0.0973550  0.1118900  0.01861800
1395  0.1

In [58]:
result2 

V1          V2          V3         V4          V5         V6         
1     0.985010    0.997030   -0.165970   0.980620    0.9517400  0.3059900 
2     0.977220    0.983680   -0.165530   0.990970    0.9623700  0.3093300 
3     0.948050    0.942440   -0.078334   0.964930    0.9912700  0.2999800 
4     0.414090    0.402880   -0.056151   0.414770    0.3990600 -0.0349970 
5     0.582510    0.574340   -0.182270   0.589240    0.6041700  0.1733100 
6     0.289360    0.280880   -0.048627   0.290170    0.2728200  0.9004400 
7     0.197800    0.210030   -0.066790   0.198570    0.1854400  0.6333200 
8     0.183230    0.213230   -0.064954   0.184010    0.1714600  0.5908600 
9     0.186980    0.216420   -0.066252   0.187760    0.1750000  0.6072600 
10    0.203670    0.197390   -0.065482   0.204390    0.1912600  0.6430300 
11    0.186980    0.216420   -0.066252   0.187760    0.1750000  0.6072600 
12    0.183230    0.213230   -0.064954   0.184010    0.1714600  0.5908600 
13   -0.070102   -0.071715    0.375760  -0.068439   -0.0751770 -0.0351290 
14    0.553630    0.544990    0.448120   0.560930    0.5796500 -0.0552040 
15    0.196470    0.190390   -0.062798   0.197180    0.1844600  0.6368800 
16    0.172850    0.219570   -0.067530   0.191460    0.1784800  0.5745100 
17    0.183230    0.213230   -0.064954   0.184010    0.1714600  0.5908600 
18   -0.106950   -0.109250    0.532680  -0.104490   -0.1143100 -0.0335180 
19    0.177960    0.216420   -0.066252   0.187760    0.1750000  0.5825300 
20    0.177960    0.216420   -0.066252   0.187760    0.1750000  0.5825300 
21    0.177960    0.216420   -0.066252   0.187760    0.1750000  0.5825300 
22    0.177960    0.216420   -0.066252   0.187760    0.1750000  0.5825300 
23    0.195780    0.216480   -0.069344   0.205710    0.1921800  0.5899400 
24    0.201390    0.213280   -0.068077   0.202170    0.1888400  0.5985800 
25    0.195780    0.216480   -0.069344   0.205710    0.1921800  0.5899400 
26   -0.083472   -0.084728    0.360640  -0.081827   -0.0878720 -0.0334980 
27   -0.042468   -0.038344   -0.044900  -0.040446   -0.0082962 -0.0067722 
28    0.600050    0.596750    0.105730   0.609540    0.6185400  0.1349300 
29    0.591750    0.588410    0.089708   0.606890    0.5725100  0.1403300 
30    0.289490    0.297330    0.121350   0.297090    0.3109000 -0.0820300 
⋮    ⋮           ⋮           ⋮          ⋮           ⋮          ⋮          
2024 -0.02617300 -0.02911200 -0.0189420 -0.02436800 -0.0338530  0.00194470
2025 -0.04218500 -0.04634000 -0.0304290 -0.03957000 -0.0421160 -0.03011500
2026 -0.03122200 -0.02890300  0.0761340 -0.02564500 -0.0247160 -0.01611800
2027 -0.04218500 -0.04634000 -0.0304290 -0.03957000 -0.0421160 -0.03011500
2028 -0.03366700 -0.03872600 -0.0175730 -0.03568000 -0.0272390 -0.05695300
2029 -0.03379400 -0.03149500  0.0744000 -0.02817000 -0.0273780 -0.01695400
2030 -0.03300700 -0.03069600  0.0717960 -0.02738600 -0.0265470 -0.01671100
2031 -0.04218500 -0.04634000 -0.0304290 -0.03957000 -0.0421160 -0.03011500
2032 -0.01060900 -0.00912980  0.0713670 -0.00950030 -0.0020405  0.00203220
2033  0.00359900  0.00034147 -0.0099817  0.00520660 -0.0044106  0.02998500
2034 -0.02505600 -0.02356400  0.0465750 -0.03736000 -0.0300180  0.00876280
2035  0.04118400  0.03997900  0.0567980  0.05053700  0.0530240  0.01178700
2036 -0.00320020 -0.00747060 -0.0277660 -0.00099723 -0.0138220 -0.01951900
2037 -0.00916030 -0.01350600 -0.0195010 -0.00684750 -0.0081100 -0.02283700
2038  0.03499500  0.02708700 -0.0135120  0.03858100  0.0237290 -0.01523100
2039  0.01247900  0.01354600 -0.0363520  0.01296000  0.0164580 -0.03235800
2040 -0.00934540 -0.01360000 -0.0154410 -0.00707610 -0.0078823 -0.02120100
2041 -0.00934540 -0.01360000 -0.0154410 -0.00707610 -0.0078823 -0.02120100
2042  0.01038900  0.00315090  0.0148820  0.00345350 -0.0059086  0.02258200
2043 -0.00637080 -0.00566920 -0.0110040 -0.00382770 -0.0074890 -0.00030993
2044 -0.00693440 -0.01132100 -0.0140630 -0.00462750 -0.0062140 -0.02362200
2045 -0.00693440 -0.01132100 -

In [56]:
result1 = fread("/home/rd2972/private_data/20240300_rss_qc_imputation/DENTIST/per_iteration/dentist_Rcpp/20240308_version/DENTIST_Rcpp_output_02.txt")

In [53]:
dentist_result = dentist(zScore = sumstat$z, LD = LD, nSample = 1153)

ERROR: Error in value[[3L]](cond): Adjusted rsq_eigen value exceeding 1: 1.002794


In [52]:
dim(dentist_LD_result)
head(dentist_LD_result)

[1] 2062 2061

LD_it   for     iteration :       V5      V6        V7        V8       
1 0.98501 0.99703 -0.165970 0.98062 0.95174  0.305990  0.196350  0.242870
2 0.97722 0.98368 -0.165530 0.99097 0.96237  0.309330  0.182120  0.245680
3 0.94805 0.94244 -0.078334 0.96493 0.99127  0.299980  0.174640  0.237840
4 0.41409 0.40288 -0.056151 0.41477 0.39906 -0.034997 -0.049078 -0.025289
5 0.58251 0.57434 -0.182270 0.58924 0.60417  0.173310  0.068187  0.168260
6 0.28936 0.28088 -0.048627 0.29017 0.27282  0.900440  0.708410 -0.039013
  V9          V10       ⋯ V2052      V2053       V2054      V2055      
1  1.9603e-01  0.200050 ⋯  0.0159500  0.01011600 -0.0094939 -1.4634e-03
2  1.9891e-01  0.184900 ⋯  0.0252130  0.01956100 -0.0130840  4.3563e-03
3  1.9103e-01  0.177670 ⋯  0.0142890  0.00809080 -0.0169470  9.1581e-05
4 -3.2538e-02 -0.046423 ⋯  0.0057649 -0.00058517 -0.0405290 -1.2314e-03
5  8.4623e-02  0.077967 ⋯ -0.0131740 -0.02404500  0.0247050 -6.2777e-03
6 -7.5352e-05  0.724340 ⋯ -0.0075276 -0.01285700 -0.0218220 -3.2347e-03
  V2056      V2057      V2058     V2059       V2060     V2061    
1 -0.0104610 -0.0104610 -0.024245 -0.00553900 -0.041944 -0.043000
2 -0.0073216 -0.0073216 -0.022075 -0.01147100 -0.039437 -0.040075
3 -0.0035575 -0.0035575 -0.027552 -0.00036976 -0.045728 -0.047427
4 -0.0481700 -0.0481700 -0.034190  0.01815900 -0.046387 -0.032131
5  0.0038312  0.0038312 -0.047474  0.04000800 -0.034784 -0.031634
6 -0.0328830 -0.0328830 -0.023340  0.04111700 -0.016986 -0.014725

Ignore the column names. The dimension of LD is not correct yet, also the diagnol elements are not equal to 1.

In [58]:
DENTIST_compile = fread("/home/rd2972/private_data/20240300_rss_qc_imputation/DENTIST/DENTIST_Rcpp_output.txt")

Warning message in fread("/home/rd2972/private_data/20240300_rss_qc_imputation/DENTIST/DENTIST_Rcpp_output.txt"):
“Detected 4 column names but the data has 2061 columns (i.e. invalid file). Added 2057 extra default column names at the end.”
Warning message in fread("/home/rd2972/private_data/20240300_rss_qc_imputation/DENTIST/DENTIST_Rcpp_output.txt"):
“Stopped early on line 2064. Expected 2061 fields but found 0. Consider fill=TRUE and comment.char=. First discarded non-empty line: <<LD_it for iteration :>>”


In [59]:
DENTIST_compile

LD_it       for         iteration  :           V5         V6         
1     0.985010    0.997030   -0.165970   0.980620    0.9517400  0.3059900 
2     0.977220    0.983680   -0.165530   0.990970    0.9623700  0.3093300 
3     0.948050    0.942440   -0.078334   0.964930    0.9912700  0.2999800 
4     0.414090    0.402880   -0.056151   0.414770    0.3990600 -0.0349970 
5     0.582510    0.574340   -0.182270   0.589240    0.6041700  0.1733100 
6     0.289360    0.280880   -0.048627   0.290170    0.2728200  0.9004400 
7     0.197800    0.210030   -0.066790   0.198570    0.1854400  0.6333200 
8     0.183230    0.213230   -0.064954   0.184010    0.1714600  0.5908600 
9     0.186980    0.216420   -0.066252   0.187760    0.1750000  0.6072600 
10    0.203670    0.197390   -0.065482   0.204390    0.1912600  0.6430300 
11    0.186980    0.216420   -0.066252   0.187760    0.1750000  0.6072600 
12    0.183230    0.213230   -0.064954   0.184010    0.1714600  0.5908600 
13   -0.070102   -0.071715    0.375760  -0.068439   -0.0751770 -0.0351290 
14    0.553630    0.544990    0.448120   0.560930    0.5796500 -0.0552040 
15    0.196470    0.190390   -0.062798   0.197180    0.1844600  0.6368800 
16    0.172850    0.219570   -0.067530   0.191460    0.1784800  0.5745100 
17    0.183230    0.213230   -0.064954   0.184010    0.1714600  0.5908600 
18   -0.106950   -0.109250    0.532680  -0.104490   -0.1143100 -0.0335180 
19    0.177960    0.216420   -0.066252   0.187760    0.1750000  0.5825300 
20    0.177960    0.216420   -0.066252   0.187760    0.1750000  0.5825300 
21    0.177960    0.216420   -0.066252   0.187760    0.1750000  0.5825300 
22    0.177960    0.216420   -0.066252   0.187760    0.1750000  0.5825300 
23    0.195780    0.216480   -0.069344   0.205710    0.1921800  0.5899400 
24    0.201390    0.213280   -0.068077   0.202170    0.1888400  0.5985800 
25    0.195780    0.216480   -0.069344   0.205710    0.1921800  0.5899400 
26   -0.083472   -0.084728    0.360640  -0.081827   -0.0878720 -0.0334980 
27   -0.042468   -0.038344   -0.044900  -0.040446   -0.0082962 -0.0067722 
28    0.600050    0.596750    0.105730   0.609540    0.6185400  0.1349300 
29    0.591750    0.588410    0.089708   0.606890    0.5725100  0.1403300 
30    0.289490    0.297330    0.121350   0.297090    0.3109000 -0.0820300 
⋮    ⋮           ⋮           ⋮          ⋮           ⋮          ⋮          
2033 -0.02617300 -0.02911200 -0.0189420 -0.02436800 -0.0338530  0.00194470
2034 -0.04218500 -0.04634000 -0.0304290 -0.03957000 -0.0421160 -0.03011500
2035 -0.03122200 -0.02890300  0.0761340 -0.02564500 -0.0247160 -0.01611800
2036 -0.04218500 -0.04634000 -0.0304290 -0.03957000 -0.0421160 -0.03011500
2037 -0.03366700 -0.03872600 -0.0175730 -0.03568000 -0.0272390 -0.05695300
2038 -0.03379400 -0.03149500  0.0744000 -0.02817000 -0.0273780 -0.01695400
2039 -0.03300700 -0.03069600  0.0717960 -0.02738600 -0.0265470 -0.01671100
2040 -0.04218500 -0.04634000 -0.0304290 -0.03957000 -0.0421160 -0.03011500
2041 -0.01060900 -0.00912980  0.0713670 -0.00950030 -0.0020405  0.00203220
2042  0.00359900  0.00034147 -0.0099817  0.00520660 -0.0044106  0.02998500
2043 -0.02505600 -0.02356400  0.0465750 -0.03736000 -0.0300180  0.00876280
2044  0.04118400  0.03997900  0.0567980  0.05053700  0.0530240  0.01178700
2045 -0.00320020 -0.00747060 -0.0277660 -0.00099723 -0.0138220 -0.01951900
2046 -0.00916030 -0.01350600 -0.0195010 -0.00684750 -0.0081100 -0.02283700
2047  0.03499500  0.02708700 -0.0135120  0.03858100  0.0237290 -0.01523100
2048  0.01247900  0.01354600 -0.0363520  0.01296000  0.0164580 -0.03235800
2049 -0.00934540 -0.01360000 -0.0154410 -0.00707610 -0.0078823 -0.02120100
2050 -0.00934540 -0.01360000 -0.0154410 -0.00707610 -0.0078823 -0.02120100
2051  0.01038900  0.00315090  0.0148820  0.00345350 -0.0059086  0.02258200
2052 -0.00637080 -0.00566920 -0.0110040 -0.00382770 -0.0074890 -0.00030993
2053 -0.00693440 -0.01132100 -0.0140630 -0.00462750 -0.0062140 -0.02362200
2054 -0.00693440 -0.01132100 -

In [28]:
kk = fread("/home/rd2972/private_data/20240300_rss_qc_imputation/DENTIST/per_iteration/dentist_original/4213_variants/LD_separate/LD_it_output_20.txt")

In [29]:
head(kk)

V1         V2          V3           V4          V5         V6        
1  0.6759160 -0.05131040 -1.92395e-03 -0.04470260  0.0505993 -0.0797282
2 -0.0286956  0.02563550  1.87835e-01 -0.03293410  0.0358906  0.0264595
3 -0.0742423  0.18604700  7.14334e-02  0.03134780 -0.0447789  0.1843570
4 -0.1909020  0.00375349 -1.68635e-03 -0.06660320  0.0394499 -0.0277484
5  0.8078360 -0.04763070 -7.06678e-05 -0.00322825  0.0854091 -0.0509672
6 -0.0531262  0.00703471  2.20129e-02  0.22299300 -0.0183365  0.0137763
  V7         V8          V9          V10         ⋯ V973       V974       
1  0.0436569  0.01908220  0.32164800  0.05074640 ⋯  0.1378230  0.11253900
2 -0.0358579 -0.04175240 -0.03144050 -0.03414390 ⋯  0.0164742  0.00802881
3 -0.1723980 -0.02332210 -0.04275870  0.00418691 ⋯ -0.0185382 -0.00656875
4  0.0172595 -0.00700894 -0.34048000 -0.05132800 ⋯ -0.0600390 -0.06478690
5  0.0533767  0.01319020  0.22029300  0.07152070 ⋯  0.0398345  0.00410843
6  0.0501553  0.02820910  0.00702352  0.07164150 ⋯ -0.0555523  0.04412450
  V975        V976        V977       V978        V979       V980      
1  0.03901070 -0.01396100 -0.0256802 -0.08519900  0.1151590 -0.2119800
2  0.00357216  0.02387510  0.0708860 -0.03225810 -0.0545845  0.0235798
3  0.01116950 -0.15475400  0.0525748  0.01008800 -0.0784397  0.0335985
4 -0.05054220 -0.00102061 -0.0511662  0.09739920 -0.0980070 -0.0827689
5  0.04491120  0.01159600 -0.0482848 -0.09297800  0.0922414 -0.2087910
6  0.09947470 -0.04475670  0.1023430  0.00352642 -0.0719310  0.0556175
  V981        V982      
1  0.04571330 -0.1430800
2 -0.00985897 -0.0141408
3 -0.07618340  0.0346303
4 -0.04171430  0.1791850
5  0.06020950 -0.1235290
6 -0.02644690 -0.0668466

In [26]:
k2 = fread("/home/rd2972/private_data/20240300_rss_qc_imputation/DENTIST/per_iteration/dentist_Rcpp/4213_variants/DENTIST_Rcpp_output_01.txt")

In [27]:
k2

V1          V2          V3         V4          V5         V6         
1     0.985010    0.997030   -0.165970   0.980620    0.9517400  0.3059900 
2     0.977220    0.983680   -0.165530   0.990970    0.9623700  0.3093300 
3     0.948050    0.942440   -0.078334   0.964930    0.9912700  0.2999800 
4     0.414090    0.402880   -0.056151   0.414770    0.3990600 -0.0349970 
5     0.582510    0.574340   -0.182270   0.589240    0.6041700  0.1733100 
6     0.289360    0.280880   -0.048627   0.290170    0.2728200  0.9004400 
7     0.197800    0.210030   -0.066790   0.198570    0.1854400  0.6333200 
8     0.183230    0.213230   -0.064954   0.184010    0.1714600  0.5908600 
9     0.186980    0.216420   -0.066252   0.187760    0.1750000  0.6072600 
10    0.203670    0.197390   -0.065482   0.204390    0.1912600  0.6430300 
11    0.186980    0.216420   -0.066252   0.187760    0.1750000  0.6072600 
12    0.183230    0.213230   -0.064954   0.184010    0.1714600  0.5908600 
13   -0.070102   -0.071715    0.375760  -0.068439   -0.0751770 -0.0351290 
14    0.553630    0.544990    0.448120   0.560930    0.5796500 -0.0552040 
15    0.196470    0.190390   -0.062798   0.197180    0.1844600  0.6368800 
16    0.172850    0.219570   -0.067530   0.191460    0.1784800  0.5745100 
17    0.183230    0.213230   -0.064954   0.184010    0.1714600  0.5908600 
18   -0.106950   -0.109250    0.532680  -0.104490   -0.1143100 -0.0335180 
19    0.177960    0.216420   -0.066252   0.187760    0.1750000  0.5825300 
20    0.177960    0.216420   -0.066252   0.187760    0.1750000  0.5825300 
21    0.177960    0.216420   -0.066252   0.187760    0.1750000  0.5825300 
22    0.177960    0.216420   -0.066252   0.187760    0.1750000  0.5825300 
23    0.195780    0.216480   -0.069344   0.205710    0.1921800  0.5899400 
24    0.201390    0.213280   -0.068077   0.202170    0.1888400  0.5985800 
25    0.195780    0.216480   -0.069344   0.205710    0.1921800  0.5899400 
26   -0.083472   -0.084728    0.360640  -0.081827   -0.0878720 -0.0334980 
27   -0.042468   -0.038344   -0.044900  -0.040446   -0.0082962 -0.0067722 
28    0.600050    0.596750    0.105730   0.609540    0.6185400  0.1349300 
29    0.591750    0.588410    0.089708   0.606890    0.5725100  0.1403300 
30    0.289490    0.297330    0.121350   0.297090    0.3109000 -0.0820300 
⋮    ⋮           ⋮           ⋮          ⋮           ⋮          ⋮          
2033 -0.02617300 -0.02911200 -0.0189420 -0.02436800 -0.0338530  0.00194470
2034 -0.04218500 -0.04634000 -0.0304290 -0.03957000 -0.0421160 -0.03011500
2035 -0.03122200 -0.02890300  0.0761340 -0.02564500 -0.0247160 -0.01611800
2036 -0.04218500 -0.04634000 -0.0304290 -0.03957000 -0.0421160 -0.03011500
2037 -0.03366700 -0.03872600 -0.0175730 -0.03568000 -0.0272390 -0.05695300
2038 -0.03379400 -0.03149500  0.0744000 -0.02817000 -0.0273780 -0.01695400
2039 -0.03300700 -0.03069600  0.0717960 -0.02738600 -0.0265470 -0.01671100
2040 -0.04218500 -0.04634000 -0.0304290 -0.03957000 -0.0421160 -0.03011500
2041 -0.01060900 -0.00912980  0.0713670 -0.00950030 -0.0020405  0.00203220
2042  0.00359900  0.00034147 -0.0099817  0.00520660 -0.0044106  0.02998500
2043 -0.02505600 -0.02356400  0.0465750 -0.03736000 -0.0300180  0.00876280
2044  0.04118400  0.03997900  0.0567980  0.05053700  0.0530240  0.01178700
2045 -0.00320020 -0.00747060 -0.0277660 -0.00099723 -0.0138220 -0.01951900
2046 -0.00916030 -0.01350600 -0.0195010 -0.00684750 -0.0081100 -0.02283700
2047  0.03499500  0.02708700 -0.0135120  0.03858100  0.0237290 -0.01523100
2048  0.01247900  0.01354600 -0.0363520  0.01296000  0.0164580 -0.03235800
2049 -0.00934540 -0.01360000 -0.0154410 -0.00707610 -0.0078823 -0.02120100
2050 -0.00934540 -0.01360000 -0.0154410 -0.00707610 -0.0078823 -0.02120100
2051  0.01038900  0.00315090  0.0148820  0.00345350 -0.0059086  0.02258200
2052 -0.00637080 -0.00566920 -0.0110040 -0.00382770 -0.0074890 -0.00030993
2053 -0.00693440 -0.01132100 -0.0140630 -0.00462750 -0.0062140 -0.02362200
2054 -0.00693440 -0.01132100 -